cpp/src/arrow/adapters/orc/adapter_util.cc

```diff
@@ -0,0 +1,422 @@
+    if (list_length && (!has_nulls || batch->notNull[i])) {
+      RETURN_NOT_OK(struct_builder->AppendValues(list_length, nullptr));
+      RETURN_NOT_OK(AdapaterUtil::AppendBatch(keytype, keys, start, list_length,
+                                              struct_builder->field_builder(0)));
+      RETURN_NOT_OK(AdapaterUtil::AppendBatch(valtype, vals, start, list_length,
+                                              struct_builder->field_builder(1)));
+    }
+  }
+  return Status::OK();
+}
+
+template <class builder_type, class batch_type, class elem_type>
+Status AppendNumericBatch(liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                          int64_t length, ArrayBuilder* abuilder) {
+  auto builder = checked_cast<builder_type*>(abuilder);
+  auto batch = checked_cast<batch_type*>(cbatch);
+
+  if (length == 0) {
+    return Status::OK();
+  }
+  const uint8_t* valid_bytes = nullptr;
+  if (batch->hasNulls) {
+    valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
+  }
+  const elem_type* source = batch->data.data() + offset;
+  RETURN_NOT_OK(builder->AppendValues(source, length, valid_bytes));
+  return Status::OK();
+}
+
+template <class builder_type, class target_type, class batch_type, class source_type>
+Status AppendNumericBatchCast(liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                              int64_t length, ArrayBuilder* abuilder) {
+  auto builder = checked_cast<builder_type*>(abuilder);
+  auto batch = checked_cast<batch_type*>(cbatch);
+
+  if (length == 0) {
+    return Status::OK();
+  }
+
+  const uint8_t* valid_bytes = nullptr;
+  if (batch->hasNulls) {
+    valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
+  }
+  const source_type* source = batch->data.data() + offset;
+  auto cast_iter = internal::MakeLazyRange(
+      [&source](int64_t index) { return static_cast<target_type>(source[index]); },
+      length);
+
+  RETURN_NOT_OK(builder->AppendValues(cast_iter.begin(), cast_iter.end(), valid_bytes));
+
+  return Status::OK();
+}
+
+Status AppendBoolBatch(liborc::ColumnVectorBatch* cbatch, int64_t offset, int64_t length,
+                       ArrayBuilder* abuilder) {
+  auto builder = checked_cast<BooleanBuilder*>(abuilder);
+  auto batch = checked_cast<liborc::LongVectorBatch*>(cbatch);
+
+  if (length == 0) {
+    return Status::OK();
+  }
+
+  const uint8_t* valid_bytes = nullptr;
+  if (batch->hasNulls) {
+    valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
+  }
+  const int64_t* source = batch->data.data() + offset;
+
+  auto cast_iter = internal::MakeLazyRange(
+      [&source](int64_t index) { return static_cast<bool>(source[index]); }, length);
+
+  RETURN_NOT_OK(builder->AppendValues(cast_iter.begin(), cast_iter.end(), valid_bytes));
+
+  return Status::OK();
+}
+
+Status AppendTimestampBatch(liborc::ColumnVectorBatch* cbatch, int64_t offset,
```

Comment:
There seems to be a lot of duplicate code here, is it possible to templatize it?

Reply:
This is just a simple code movement from original adapter.cc to here. I haven't touch any of the conversion logic yet. Would prefer to perform this part of work in separate PR.

Reply:
SGTM